docs/reference/modules/coordination.asciidoc

```diff
@@ -0,0 +1,461 @@
+
+If it is not possible to use this tool, you can also bootstrap the cluster via
+the API as described here. There are two steps to the bootstrapping process.
+Firstly, after all the nodes have started up, created their persistent node
+IDs, and discovered each other, the first step is to request a bootstrap
+document:
+
+[source,js]
+--------------------------------------------------
+# Return the current bootstrap document immediately
+GET /_cluster/bootstrap
+# Wait until the node has discovered at least 3 nodes, or 60 seconds has elapsed,
+# and then return the bootstrap document
+GET /_cluster/bootstrap?wait_for_nodes=3&timeout=60s
+--------------------------------------------------
+// CONSOLE
+
+The boostrap document contains information that the cluster needs to start up,
+and looks like the following.
+
+[source,js]
+--------------------------------------------------
+{
+  "master_nodes_failure_tolerance": 1,
+  "master_nodes":[
+    {"id":"USpTGYaBSIKbgSUJR2Z9lg"},
+    {"id":"gSUJR2Z9lgUSpTGYaBSIKb"},
+    {"id":"2Z9lgUSpTgSUYaBSIKbJRG"}
+  ]
+}
+--------------------------------------------------
+
+It is safe to repeatedly call `GET /_cluster/bootstrap`, and to call it on
+different nodes concurrently. This API will yield an error if the receiving
+node has already been bootstrapped or has joined an existing cluster.
+
+Once a bootstrap document has been received, it must then be sent back to the
+cluster to finish the bootstrapping process as follows:
+
+[source,js]
+--------------------------------------------------
+# send the bootstrap document back to the cluster
+POST /_cluster/bootstrap
+{
+  "master_nodes_failure_tolerance": 1,
+  "master_nodes":[
+    {"id":"USpTGYaBSIKbgSUJR2Z9lg"},
+    {"id":"gSUJR2Z9lgUSpTGYaBSIKb"},
+    {"id":"2Z9lgUSpTgSUYaBSIKbJRG"}
+  ]
+}
+--------------------------------------------------
+// CONSOLE
+
+It is safe to repeatedly call `POST /_cluster/bootstrap`, and to call it on
+different nodes concurrently, but **it is vitally important** to use the same
+bootstrap document in each call.
+
+It is also possible to select the initial set of nodes in terms of their names
+rather than their IDs as follows.
+
+[source,js]
+--------------------------------------------------
+# send the bootstrap document back to the cluster
+POST /_cluster/bootstrap
+{
+  "master_nodes_failure_tolerance": 1,
+  "master_nodes":[
+    {"name":"master-a"},
+    {"name":"master-b"},
+    {"name":"master-c"}
+  ]
+}
+--------------------------------------------------
+// CONSOLE
+
+This can be useful if the node names are known (and known to be unique) in
+advance, and means that the first `GET /_cluster/bootstrap` call is not
+necessary. As above, it is safe to repeatedly call `POST /_cluster/bootstrap`,
+and to call it on different nodes concurrently, but **it is vitally important**
+to use the same bootstrap document in each call.
+
+[float]
+=== Manually-triggered elections
```

Comment:
I wonder if we should leave this chapter out for now. The chapter does not say why you would manually trigger an election. If we don't have a use-case for this, it's probably better to leave it out.

Reply:
@DaveCTurner I agree with @ywelsch that we need to give a use case to the user why manual elections could be useful, especially if the manual election is "best-effort"

Reply:
I don't think that #14340 requires manually-triggered elections. I think that that issue could be solved in a fully automated way by handing over leadership to nodes that are not marked as "voting-only". Again, I don't think we should have this in the docs before we have built a feature for it.

Reply:
It doesn't _require_ it, sure, we could achieve the same end in other ways. The ability to pick a different master also allows for a multi-tenant setup to move their elected masters around to balance the master-related load. We could also achieve this in other ways, I'm sure.

The intention of this document was not to describe the API that will be in the first cut, and I wouldn't be surprised if this particular endpoint isn't implemented until later, but I think we should consider it in the design. It's very cheap to implement this as described.

Reply:
it seems this discussion goes beyond docs - do we have any reason to build these APIs?

Reply:
I don't understand your question @boaz. There are two reasons mentioned above:

1. #14340 asks for the ability to switch to a different master (possibly automatically)
2. In a multi-tenant environment, avoid having all the masters be too concentrated.

I'm :neutral_face: on thinking about the design for this now. I'm ok with ignoring it until we start to think about implementing these features (maybe never) and hope that we can sensibly fit them into whatever we decided for the rest of the API.

Reply:
@DaveCTurner that was my point - this is a discussion about more that just having docs but we're also seem to be designing APIs for those use cases (but do not need them for other functionality). I think we can leave them out of scope for now.

Reply:
I have removed this API from the document.